### PR TITLE
[WIP] Introduce `List.keytake/3`

### PR DESCRIPTION
**Pull request by lexmag**

### CODE REVIEW

File: lib/elixir/lib/list.ex [hunk 1]
```diff
@@ -275,6 +275,26 @@ defmodule List do
...
+      {:value, {:a, 1}, [b: 2]}
+
+      iex> List.keytake([a: 1, b: 2], 2, 1)
+      {:value, {:b, 2}, [a: 1]}
+
+      iex> List.keytake([a: 1, b: 2], :c, 0)
+      false
+
+  """
+  @spec keytake([tuple], any, non_neg_integer) :: {:value, tuple, [tuple]} | false
```
Comment (lexmag) [inline lib/elixir/lib/list.ex]: I think we can improve the function return, make it either `{any, [tuple]}` or `{tuple, [tuple]} | nil`. @josevalim what are your thoughts?

Comment (josevalim) [inline lib/elixir/lib/list.ex]: I like `{tuple, [tuple]} | nil`.

Comment (lexmag) [inline lib/elixir/lib/list.ex]: Changed to `{tuple, [tuple]} | nil` (which is more alike to original return). How do you feel about adding `keypop/4` (`{any, [tuple]}` and defaults) to have something similar to `Dict.pop/3`?

**Comment by eksperimental** [comment]
I think the convention should be to name this function: List.key_take/3

**Comment by lexmag** [comment]
@eksperimental the convention for `List.key*` functions is formed, e.g. (https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/list.ex#L214).

**Comment by eksperimental** [comment]
@lexmag true. I think those are breaking the rule as well, I was going to send a proposal to replace a few that i have found, I just haven't had the time.

**Comment by alco** [comment]
By Elixir convention they'd be called `replace_key`, `take_key` etc. But those `List.key*` functions have been inherited from Erlang, changing them now would add more confusion that do good. It's the culture baggage of Elixir, every language has those :)

**Comment by eksperimental** [comment]
@alco, I know, but i have noticed other places where the elixir functions have been named same as the erlang counterpart. I think it's something to take into account when creating new functions.

how difficult would be to rename them?

**Comment by ericmj** [comment]
@eksperimental We can't rename them, at least not until 2.0 which is still far away. What other places are you referring to?

**Comment by eksperimental** [comment]
@ericmj I will send them to the core mailing list, I had them in a file somewhere. i will do that as soon as I get hold of that file.

**Comment by lexmag** [comment]
@alco well said. :smile:

**Comment by bcardarella** [comment]
I hate to be "that guy" but how come this method doesn't have a test associated with it? Shouldn't all new public API get that?

**Comment by lexmag** [comment]
@bcardarella this function is covered via doctests.

**Comment by josevalim** [comment]
@bcardarella and please be "that guy" if it means pointing out some code has no tests. :)

**Pull request merged**